### PR TITLE
Add documentation for an overview of the goto program transformation passes

### PR DESCRIPTION
**Pull request by thomasspriggs**
This PR adds documentation for an overview of the goto program transformation passes. I'd prefer to get an initial version of this documentation merged with the level of detail which is currently in this branch. We can then keep track of requests for further detail using issues such as https://github.com/diffblue/cbmc/issues/7546 if more detail is needed.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by TGWDB** [comment]
Fixes #7546

File: doc/architectural/goto-program-transformations.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+[Linking of Standard Libraries]._
+
+### Removal/Lowering of Function Pointers
+
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that match the type of the function pointer.
+
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: It does not seem to work like this in practice. Targets are identified by matching signature arity, and then each target gets typecasted arguments

File: doc/architectural/goto-program-transformations.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+`other` instruction.
+
+_This pass has no predeccesor._
+
+### String Instrumentation
+
+This (optional) pass adds checks on calls to some of the C standard library
+string functions. It uses the tracking information added by the
+[string-abstraction] pass. It is switched on by the `--string-abstraction`
+command line option.
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: Do we have a description of what string abstraction does ? In particular once applied it does not allow to analyze properties related to the contents of the strings since it only tracks string lenght.

File: doc/architectural/goto-program-transformations.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+### Removal/Lowering of Function Pointers
+
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that match the type of the function pointer.
+
+Note that this pass must be applied after all linking has been done. Either for
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: Is there any way to enforce or check this requirement ?
It seems possible to force pointer removal using a goto-instrument call on a model that has not yet been full linked or linked to the CPROVER library.

File: doc/architectural/goto-program-transformations.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+See [modeling-mmio.md] for details of modeling memory-mapped I/O regions of
+device interfaces. This pass is always carried out but will only make changes if
+one of the modelling functions exist.
+
+_Predecessor pass is [Removal/Lowering of Function Pointers]._
+
+
+### Instrument/Remove Preconditions
+
+This pass moves preconditions associated with functions to call sites rather
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md:74]: Could we add a short example of what the transformation does ?

File: doc/architectural/goto-program-transformations.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+sites.
+
+_Predecessor pass is [Memory Mapped IO Instrumentation]._
+
+
+### Removal/Lowering of Return Statements
+
+This pass replaces returns of values with writes and reads to global variables.
+
+_Predecessor passes are [Instrument/Remove Preconditions] or [Partial Inlining]._
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: Can we add a code snippet that shows how the transformation works ? Are there any constraints on the names of these variables ? (taking the perspective of someone who would like to produce such a model from scratch)

File: doc/architectural/goto-program-transformations.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+- location numbers
+- loop numbers
+
+_Predecessor passes are [Adjust Float Expressions] or [String Abstraction]._
+
+
+### Add Failed Symbols
+
+This pass adds failed symbols to the symbol table only. See
+`src/pointer-analysis/add_failed_symbols.h` for details.
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: What is a failed symbol, why/how is it used? the source code does not contain a high-level description of this.

File: doc/architectural/goto-program-transformations.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+_Predecessor passes are [Goto Functions Update] or
+[Add Non-Deterministic Initialisation of Global Scoped Variables]._
+
+
+### Remove Skip Instructions
+
+This transformation removes skip instructions. Note that this transformation is
+called in many places and may be called as part of other transformations. This
+instance here is part of the mandatory transformation to reach core goto format.
+
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: So does this mean that a GOTO program with SKIP instructions is not in core format ?

File: doc/architectural/goto-program-transformations.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+### Remove Unused Functions
+
+This pass removes unused functions from the goto model. In practice this builds
+a collection of all the functions that are potentially called, and then removes
+any function not in this collection.
+
+_Predecessor pass is [Add Failed Symbols]._
+
```
Comment (remi-delmas-3000) [inline doc/architectural/goto-program-transformations.md]: This pass triggers function pointer removal when applied using goto-instrument, but may not trigger it when applied programatically. Is this pass sound when all pointers have not been resolved ? Should it have function pointer removal as a precondition ?

**Review by remi-delmas-3000 — DISMISSED** [review]

**Comment by remi-delmas-3000** [comment]
Looking at the core GOTO definition and this list of transformation, they do not seem to discuss entry points,  or support function generation for __CPROVER_start, __CPROVER_initialize, or the initialisation of statics and their restrictions (e.g. no function calls in __CPROVER_initialize). What would be a good place to describe these aspects ?

File: doc/architectural/goto-program-transformations.md (hunk 7 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Support for `SKIP` instructions is implemented in symex. The support should be considered to be untested given that the instructions are routinely removed before symex; not that a `SKIP` instruction is particularly complex to deal with. I would say that the ideal should be not to include them in the core format for maximum certainty that the analysis will be using well tested code. YMMV

File: doc/architectural/goto-program-transformations.md (hunk 3 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: The requirement is checked by an assertion that the fall through is unreachable. See - https://github.com/diffblue/cbmc/blob/d7099aeaff5043b65aba669fe221cc9b9af40c8c/src/goto-programs/remove_function_pointers.cpp#L435

I may have worded this a little too strongly. If you can show via data flow analysis that all potential targets of a function pointer at the point it is called are included in the model then applying this pass to that function call would yield a good result. Indeed this could result in a faster analysis due to the smaller set of functions being included in the switch block.

The current approach this pass uses has the drawback that the more functions which are linked into the model the larger the function table generated may be, but the earlier it is done in terms of linking then the more chance that valid targets are not included. Resolving this would require rewriting it using an entirely different approach. Two  potential solutions I can think of would be -

1. Make the lowering part of symex. This would be a fairly complete reworking of the current implementation and would add to the complexities of symex.
2. Use the Abstract Interpreter with variable sensitivity to build the value sets of function pointers. This would require fixing the known issues which the current implementation has with recursive function calls.

File: doc/architectural/goto-program-transformations.md (hunk 1 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: I guess this should should say "loosely match". The number of arguments is checked. As I read the `arg_is_type_compatible` function, cases where the types are compound types such as structs or arrays and the number of bits of the two types are known to be different will be excluded.

File: doc/architectural/goto-program-transformations.md (hunk 6 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: They are explained here - https://github.com/diffblue/cbmc/blob/d7099aeaff5043b65aba669fe221cc9b9af40c8c/src/goto-symex/symex_dereference.cpp#L457

My reading of this is that these symbols are used as an (undefined) resulting value of a pointer dereference in the case where a pointer may point to a target which is not defined by the model.

File: doc/architectural/goto-program-transformations.md (hunk 5 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: The names end in `#return_value` as is defined by the `RETURN_VALUE_SUFFIX` macro. They are the combination of the the functions' base_name and this suffix.

The snippet already exists in the documentation at the top of `src/goto-programs/remove_returns.h`. I am planning to add a link to this, before I move the PR out of draft status.

**Comment by thomasspriggs** [comment]
> Looking at the core GOTO definition and this list of transformation, they do not seem to discuss entry points, or support function generation for __CPROVER_start, __CPROVER_initialize, or the initialisation of statics and their restrictions (e.g. no function calls in __CPROVER_initialize). What would be a good place to describe these aspects ?

These functions are initially generated in the symbol table using a source language specific implementation. They are then translated into GOTO using a source language specific conversion. The existence of the `__CPROVER_start` would be a core goto requirement, but its creation is not a goto to goto transformation.

File: doc/architectural/goto-program-transformations.md (hunk 8 above)
Comment (jimgrundy) [reply doc/architectural/goto-program-transformations.md]: +1

File: doc/architectural/goto-program-transformations.md [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+_Predecessor passes are [Remove Skip Instructions] or [Add Coverage Goals]._
+
+
+
+## Optional Transformation Passes
+
+The sections lists the optional transformation passes that are optional and will
+modify a goto model. Note that these are documented here for consistency, but
+not require for core goto format.
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md]: "not require" -> "not required"

File: doc/architectural/goto-program-transformations.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+Note for each optional pass there is a listed predeceesor pass. This is the pass
+currently called before the listed pass in CBMC. While the ordering may not be
+fully documented, this should be followed.
+
+### Linking of Standard Libraries
+
+This pass links the function defenitions from the Cprover standard library
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md]: "defenitions" spelling.

File: doc/architectural/goto-program-transformations.md (hunk 1 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Updated.

File: doc/architectural/goto-program-transformations.md (hunk 9 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Done.

File: doc/architectural/goto-program-transformations.md (hunk 10 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Fixed.

File: doc/architectural/goto-program-transformations.md (hunk 8 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Adding (or improving) a precondition would be out-of-scope for a documentation PR. The remove unused functions pass makes the assumption that `function_call` instructions have their `call_function` value assigned to a symbol expression. This assumption is already checked inside the `to_symbol_expr` function used by the pass.

Calls to function pointers which have not been lowered are `function_call` instructions where the value of `call_function` is a dereference expression. Therefore I would expect that programmatically calling the remove unused functions pass on a program which contains an unlowered function pointer call will result in an invariant violation; thus preventing any kind of analysis sound or otherwise from taking place.

File: doc/architectural/goto-program-transformations.md (hunk 2 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: There is a description of what it does associated with the `string_abstractiont` class. My understanding is that it only adds tracking information rather than discarding any information. So there are no properties which you can't analyse when the pass is in place, which you can analyse when it isn't. Tracking the length simply helps improve the efficiency of analysing length based assertions because it avoids having to iterate through C strings to work out how long they are as part of the assertion.

**Comment by thomasspriggs** [comment]
I'd prefer to get an initial version of this documentation merged with the level of detail which is currently in this branch. We can then keep track of requests for further detail using issues such as https://github.com/diffblue/cbmc/issues/7546 if more detail is needed.

I am happy to fix any spelling/grammatical mistakes needed to get this merged. If there is disagreement about what a particular pass actually does, then I propose provisionally removing the contentious details from this PR leaving references to the implementation instead.

**Review by TGWDB — APPROVED** [review]
There may be some minor/subtle changes advisable in future, but this is an excellent start and description of the passes and their partial ordering that should be merged.

File: doc/architectural/goto-program-transformations.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+in \ref remove_function_pointers.h
+
+<em>Predecessor pass is the \ref linking-transform or the optional
+\ref string-instrument-transform if it is being used.</em>
+
+\subsection mmio-transform Memory Mapped IO Instrumentation
+
+This pass finds pointer dereferences and adds corresponding calls to the
+`__CPROVER_mm_io_r` and `__CPROVER_mm_io_w` modelling functions if they exist.
+See [modeling-mmio.md](Device-behavior) for details of
```
Comment (NlightNFotis) [inline doc/architectural/goto-program-transformations.md]: For the markdown linking to work, you need the link inside the parentheses:

```
[Device-behavior](modelling-mmio.md)

File: doc/architectural/goto-program-transformations.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+<em>Predecessor pass is \ref remove-skip-transform .</em>
+
+
+\subsection slicing-transforms Slicing
+
+This includes several slicing passes as specified by various slicing command
+line options. The reachability slicer is enabled by the `--reachability-slice`
+command line option. The implementation of this pass is called via the \ref
+reachability_slicer(goto_modelt &, message_handlert &) function. The full slicer
+is enabled by the `--full-slice` command line option. The implementation of this
```
Comment (NlightNFotis) [inline doc/architectural/goto-program-transformations.md:312]: Should we add some links to the documents that mentioned some of these options were unsound? https://github.com/diffblue/cbmc/pull/6479/files

File: doc/architectural/goto-program-transformations.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+string-abstraction-transform if it is being used.</em>
+
+
+\subsection failed-symbols-transform Add Failed Symbols
+
+This pass adds failed symbols to the symbol table. See
+`src/pointer-analysis/add_failed_symbols.h` for details. The implementation of
+this pass is called via \ref add_failed_symbols(symbol_table_baset &) . The
+purpose of failed symbols is explained in the documentation of the function \ref
+goto_symext::dereference(exprt &, statet &, bool)
```
Comment (NlightNFotis) [inline doc/architectural/goto-program-transformations.md:177]: This hasn't managed to link accurately.

**Review by NlightNFotis — APPROVED** [review]
Approved with minor comments.

File: doc/architectural/goto-program-transformations.md (hunk 13 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:177]: This link actually works when I render it on my machine.

File: doc/architectural/goto-program-transformations.md (hunk 12 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:312]: Might be nice to add on a follow-up PR.

File: doc/architectural/goto-program-transformations.md [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+\page goto-program-transformations Goto Program Transformations
+
+\section required-transforms Core Transformation Passes
+
+This section lists the transformation passes that must have been applied for the
+goto model to be in core goto format.
+
+Note that the passes are listed below in the order they are currently called in
+CBMC. While all dependencies on the ordering are not fully documented, the
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md:11]: Can I read that as "Until all dependencies are fully documented and checked, the following order should be used."?

File: doc/architectural/goto-program-transformations.md [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that approximately match the type of the function pointer.
+
+Note that for this pass to work as intended, all potential targets of calls to
+function pointers must be in the model. Otherwise they will not be added to the
+relevant switch blocks. This may cause the assertion that the switch blocks'
+fallthrough is unreachable to be shown to be violated via an invalid counter
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md:46]: Less confusing to me to just say "via a counter example" rather than an "invalid counter example".

File: doc/architectural/goto-program-transformations.md (hunk 4 above)
Comment (jimgrundy) [reply doc/architectural/goto-program-transformations.md:74]: Is it a weird asymmetry that there isn't something analogous that moves the post-condition checks to the call site rather than the tail of the function? Or is there such a beast and I am just missing the reference to it?

File: doc/architectural/goto-program-transformations.md [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+The implementation of this pass is called via \ref remove_skip(goto_modelt &)
+
+<em>Predecessor pass is \ref failed-symbols-transform or the optional \ref
+unused-functions-transform if it is being used.</em>
+
+
+
+\subsection properties-transform Label Properties
+
+This transformation adds information to assert instructions. This includes
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md]: The need for this transform is troubling. How are we letting in assertions that aren't annotated with source information? I think the right solution is to start forcing code to have locations when it is added.

File: doc/architectural/goto-program-transformations.md [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+variables including static variables. For details see
+`src/goto-instrument/nondet_static.h`. The initialisation code is added to the
+`CPROVER_initialize` function in the goto model. The implementation of this pass
+is called via the \ref nondet_static(goto_modelt &) function.
+
+<em>Predecessor pass is \ref update-transform.</em>
+
+
+\subsection unused-functions-transform Remove Unused Functions
+
```
Comment (jimgrundy) [inline doc/architectural/goto-program-transformations.md:280]: Is it important not to do this until after the transformation that replaces calls via function pointers into case-statements? I could imagine a function getting thrown away and then not being considered as a possible function pointer target.

File: doc/architectural/goto-program-transformations.md (hunk 14 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:11]: You could. I think it is also important to recognise that the ordering in this documentation is well tested, in comparison to any alternative ordering of passes. Therefore the guarantees of proper functioning for alternative orders will be weaker, even if pass inter-dependencies were to be well documented.

File: doc/architectural/goto-program-transformations.md (hunk 15 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:46]: :thinking: The counter example would be correct in terms of the model, but incorrect compared to a concrete execution. The counter example states a way of reaching a fall though which does not exist in the source program. So a concrete execution matching the trace would result in a real function (which was removed from the model) being called instead of the generated fall-through.

File: doc/architectural/goto-program-transformations.md (hunk 16 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: On further inspection I think that the main purpose is to assign unique ids to properties. Unique IDs are not assigned in the source language so this needs to be done somewhere and this pass is where it is done. Adding source locations is fall-back functionality which should not be needed. I assume it is there as we can't tag the ID onto the location if the location is missing. My description here in this documentation is based on my inspection of the implementation...

We could try adding an INVARIANT into the pass which checks that the location already exists and remove the code which adds locations. However my aim when writing this documentation was to document the current state of the code base rather than the ideal state of it, which would be misleading. I would argue that it is preferable to merge a PR documenting the current state, even if that state is undesirable. This helps with planning worthwhile follow-up work. The documentation then needs to be updated in tandem with improvements to the implementations of the transformation passes.

File: doc/architectural/goto-program-transformations.md (hunk 17 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:280]: I think I responded to this observation previously. It is important to remove calls via function pointers before this pass because the remove unused functions pass will violate an INVARIANT if you do not.

File: doc/architectural/goto-program-transformations.md (hunk 4 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md:74]: I am not sure. @tautschnig I think you may be more familiar with the contracts work than me, though I think this pass may intended for models of library functions, rather than contracts - Can you share any insights into why we need to do this with preconditions, but not postconditions?

File: doc/architectural/goto-program-transformations.md (hunk 14 above)
Comment (jimgrundy) [reply doc/architectural/goto-program-transformations.md:11]: Sure. There is what works and what is expected to work. At some point we would expect things to work if you met the documented dependencies, and if they didn't that would be an issue to get fixed.

File: doc/architectural/goto-program-transformations.md (hunk 16 above)
Comment (jimgrundy) [reply doc/architectural/goto-program-transformations.md]: So can we just reword this to emphasize what is expected? Maybe:

> This transformation adds unique IDs to the location information for a property. It will also infer and populate file and line information from the surrounding context if this is missing.

File: doc/architectural/goto-program-transformations.md (hunk 17 above)
Comment (jimgrundy) [reply doc/architectural/goto-program-transformations.md:280]: Sure, but one might argue then if you've had to respond to this before then kind of makes the point that readers could use a reminder. Maybe something like:
> **Note:** Removing unused functions before removing/lowering function pointers may result in false counter-examples.

File: doc/architectural/goto-program-transformations.md (hunk 16 above)
Comment (thomasspriggs) [reply doc/architectural/goto-program-transformations.md]: Ok. I think I am going to remove any reference to location addition from this section. The pass doesn't infer any file/line information. It just adds the blank data structure, so that the property IDs can be added. The file, line, ~~function,~~ column etc. members of the location structure will remain empty. The location structure addition is just an implementation detail of adding the property identifier in this case.
